ext/cgi/escape/escape.c

```diff
@@ -44,36 +44,28 @@ preserve_original_state(VALUE orig, VALUE dest)
 static VALUE
 optimized_escape_html(VALUE str)
 {
-    long i, len, beg = 0;
-    VALUE dest = 0;
-    const char *cstr;
-
-    len  = RSTRING_LEN(str);
-    cstr = RSTRING_PTR(str);
-
-    for (i = 0; i < len; i++) {
-	switch (cstr[i]) {
-	  case '\'':
-	  case '&':
-	  case '"':
-	  case '<':
-	  case '>':
-	    if (!dest) {
-		dest = rb_str_buf_new(len);
-	    }
-
-	    rb_str_cat(dest, cstr + beg, i - beg);
-	    beg = i + 1;
-
-	    html_escaped_cat(dest, cstr[i]);
-	    break;
-	}
+    const char *cstr = RSTRING_PTR(str);
+    const char *end = cstr + RSTRING_LEN(str);
+    char *buf = ALLOCA_N(char, RSTRING_LEN(str) * html_escape_max_len + 1);
+
+    char *dest = buf;
+    while (cstr < end) {
+        const unsigned char c = *cstr++;
+        const char *escaped = html_escape_table[c];
+        if (escaped) {
+            memcpy(dest, escaped, html_escape_len[c]);
+            dest += html_escape_len[c];
+        }
+        else {
+            *dest++ = c;
+        }
     }
 
-    if (dest) {
-	rb_str_cat(dest, cstr + beg, len - beg);
-	preserve_original_state(str, dest);
-	return dest;
+    if (RSTRING_LEN(str) < (dest - buf)) {
+        *dest = '\0';
+        VALUE escaped = rb_str_new_cstr(buf);
```

Comment:
As `rb_str_new_cstr` calls `strlen`, `VALUE escaped = rb_str_new(buf, dest - buf);` would be faster.

Reply:
cool, thanks https://github.com/ruby/ruby/pull/2226/commits/751b4bf1b5ac8799a581d79524b72a4d11e33a44